rust/src/detect/transform_base64.rs

```diff
@@ -0,0 +1,432 @@
+
+pub const DETECT_TRANSFORM_BASE64_MIN_PARAM_COUNT: usize = 1;
+pub const DETECT_TRANSFORM_BASE64_MAX_PARAM_COUNT: usize = 3;
+pub const DETECT_TRANSFORM_BASE64_FLAG_MODE: u8 = 0x01;
+pub const DETECT_TRANSFORM_BASE64_FLAG_NBYTES: u8 = 0x02;
+pub const DETECT_TRANSFORM_BASE64_FLAG_OFFSET: u8 = 0x04;
+pub const DETECT_TRANSFORM_BASE64_FLAG_OFFSET_VAR: u8 = 0x08;
+pub const DETECT_TRANSFORM_BASE64_FLAG_NBYTES_VAR: u8 = 0x10;
+
+#[repr(C)]
+#[derive(Debug)]
+pub struct DetectTransformFromBase64Data {
+    flags: u8,
+    nbytes: u32,
+    nbytes_str: *const c_char,
+    offset: u32,
+    offset_str: *const c_char,
+    mode: DetectBase64Mode,
+}
+
+impl Drop for DetectTransformFromBase64Data {
+    fn drop(&mut self) {
+        unsafe {
+            if !self.offset_str.is_null() {
+                let _ = CString::from_raw(self.offset_str as *mut c_char);
+            }
+            if !self.nbytes_str.is_null() {
+                let _ = CString::from_raw(self.nbytes_str as *mut c_char);
+            }
+        }
+    }
+}
+impl Default for DetectTransformFromBase64Data {
+    fn default() -> Self {
+        DetectTransformFromBase64Data {
+            flags: 0,
+            nbytes: 0,
+            nbytes_str: std::ptr::null_mut(),
+            offset: 0,
+            offset_str: std::ptr::null_mut(),
+            mode: TRANSFORM_FROM_BASE64_MODE_DEFAULT,
+        }
+    }
+}
+
+impl DetectTransformFromBase64Data {
+    pub fn new() -> Self {
+        Self {
+            ..Default::default()
+        }
+    }
+}
+
+fn get_mode_value(value: &str) -> Result<DetectBase64Mode, ()> {
+    let res = match value {
+        "relax" => DetectBase64Mode::ModeRelax,
+        "rfc4648" => DetectBase64Mode::ModeRFC4648,
+        "rfc2045" => DetectBase64Mode::ModeRFC2045,
+        "strict" => DetectBase64Mode::ModeStrict,
+        _ => return Err(()),
+    };
+
+    Ok(res)
+}
+
+fn parse_transform_base64(
+    input: &str,
+) -> IResult<&str, DetectTransformFromBase64Data, RuleParseError<&str>> {
+    // Inner utility function for easy error creation.
+    fn make_error(reason: String) -> nom7::Err<RuleParseError<&'static str>> {
+        Err::Error(RuleParseError::InvalidTransformBase64(reason))
+    }
+    let mut transform_base64 = DetectTransformFromBase64Data::new();
+
+    if input.is_empty() {
+        return Ok((input, transform_base64));
+    }
+    let (_, values) = nom7::multi::separated_list1(
+        tag(","),
+        preceded(multispace0, nom7::bytes::complete::is_not(",")),
+    )(input)?;
+
+    // No options so return defaults
+    if values.is_empty() {
+        return Ok((input, transform_base64));
+    };
+
+    if values.len() < DETECT_TRANSFORM_BASE64_MIN_PARAM_COUNT
+        || values.len() > DETECT_TRANSFORM_BASE64_MAX_PARAM_COUNT
+    {
+        return Err(make_error(format!("Incorrect argument string; at least {} values must be specified but no more than {}: {:?}",
```

Comment:
is this rustfmt clean? Seems like a very long line. For a new file it seems we can enforce being clean

Reply:
Yes, `rustfmt` is clean.